File: packages/core/benchmarks/performance.ts
/**
 * Performance benchmarks for TypeScript fallback implementations.
 *
 * Run with: cd packages/core && bun run benchmarks/performance.ts
 *
 * This benchmarks the pure TypeScript implementations to establish
 * a baseline for comparison with Rust native bindings.
 */

import { Bench } from 'tinybench'
import {
  chunkTextFallback,
  cosineSimilarityFallback,
  hybridMergeResultsFallback,
  normalizeVectorFallback,
  vectorDistanceFallback,
  type VectorResult,
} from '../src/memory.js'

// ============================================================================
// Test Data Generation
// ============================================================================

function generateVector(size: number): number[] {
  return Array.from({ length: size }, (_, i) => Math.sin(i / size))
}

function generateMarkdownDocument(sections: number): string {
  let doc = '# Document Title\n\nIntroduction paragraph with some content.\n\n'

  for (let i = 0; i < sections; i++) {
    doc += `## Section ${i + 1}\n\n`
    doc += 'This is a paragraph of text that contains some content. '
    doc += 'It has multiple sentences to simulate real documentation. '
    doc += `This is section number ${i + 1}.\n\n`

    // Add code block occasionally
    if (i % 5 === 0) {
      doc += '```rust\nfn example() -> u32 {\n    42\n}\n```\n\n'
    }

    // Add list occasionally
    if (i % 3 === 0) {
      doc += '- Item one\n- Item two\n- Item three\n\n'
    }
  }

  return doc
}

function generateVectorResults(size: number): VectorResult[] {
  return Array.from({ length: size }, (_, i) => ({
    id: `doc_${i}`,
    score: 0.95 - i * 0.0001,
  }))
}

function generateKeywordResults(size: number): VectorResult[] {
  return Array.from({ length: size }, (_, i) => ({
    id: i % 2 === 0 ? `doc_${i}` : `kw_doc_${i}`,
    score: 15.0 - i * 0.01,
  }))
}

// ============================================================================
// Benchmarks
// ============================================================================

async function runBenchmarks() {
  console.log('🚀 TypeScript Performance Benchmarks\n')
  console.log('=' .repeat(70))

  // ----- Cosine Similarity Benchmarks -----
  console.log('\n📊 Cosine Similarity Benchmarks\n')

  const cosineBench = new Bench({ time: 1000 })

  for (const size of [128, 512, 1024, 1536, 4096]) {
    const a = generateVector(size)
    const b = Array.from({ length: size }, (_, i) => Math.cos(i / size))

    cosineBench.add(`cosine_similarity_dim_${size}`, () => {
      cosineSimilarityFallback(a, b)
    })
  }

  await cosineBench.warmup()
  await cosineBench.run()
  console.table(cosineBench.table())

  // ----- Chunk Markdown Benchmarks -----
  console.log('\n📊 Chunk Markdown Benchmarks\n')

  const chunkBench = new Bench({ time: 1000 })

  const smallDoc = '# Title\n\nShort content paragraph here.\n\n## Section\n\nMore text.'
  const mediumDoc = generateMarkdownDocument(50)  // ~10KB
  const largeDoc = generateMarkdownDocument(500)  // ~100KB

  console.log(`Document sizes: small=${smallDoc.length}b, medium=${mediumDoc.length}b, large=${largeDoc.length}b`)

  chunkBench.add('chunk_small_100b', () => chunkTextFallback(smallDoc, 512))
  chunkBench.add('chunk_medium_10kb', () => chunkTextFallback(mediumDoc, 512))
  chunkBench.add('chunk_large_100kb', () => chunkTextFallback(largeDoc, 512))
  chunkBench.add('chunk_medium_256tok', () => chunkTextFallback(mediumDoc, 256))
  chunkBench.add('chunk_medium_1024tok', () => chunkTextFallback(mediumDoc, 1024))

  await chunkBench.warmup()
  await chunkBench.run()
  console.table(chunkBench.table())

  // ----- Hybrid Merge Benchmarks -----
  console.log('\n📊 Hybrid Merge Benchmarks\n')

  const hybridBench = new Bench({ time: 1000 })

  for (const size of [50, 100, 500, 1000, 5000]) {
    const vecResults = generateVectorResults(size)
    const kwResults = generateKeywordResults(Math.floor(size / 2))

    hybridBench.add(`hybrid_merge_results_${size}`, () => {
      hybridMergeResultsFallback(vecResults, kwResults, 0.7, 0.3, 50)
    })
  }

  await hybridBench.warmup()
  await hybridBench.run()
  console.table(hybridBench.table())

  // ----- Vector Operations Benchmarks -----
  console.log('\n📊 Vector Operations Benchmarks\n')

  const vectorBench = new Bench({ time: 1000 })

  const vec1024 = generateVector(1024)
  const vec1536 = generateVector(1536)

  vectorBench.add('normalize_1024', () => normalizeVectorFallback(vec1024))
  vectorBench.add('normalize_1536', () => normalizeVectorFallback(vec1536))
  vectorBench.add('distance_1024', () => vectorDistanceFallback(vec1024, vec1024))
  vectorBench.add('distance_1536', () => vectorDistanceFallback(vec1536, vec1536))

  await vectorBench.warmup()
  await vectorBench.run()
  console.table(vectorBench.table())

  // ----- Edge Cases -----
  console.log('\n📊 Edge Case Benchmarks\n')

  const edgeBench = new Bench({ time: 500 })

  const emptyVec: number[] = []
  const identicalVec = Array.from({ length: 1024 }, (_, i) => i)
  const longLine = 'word '.repeat(10000)

  edgeBench.add('cosine_empty', () => cosineSimilarityFallback(emptyVec, emptyVec))
  edgeBench.add('cosine_identical_1024', () => cosineSimilarityFallback(identicalVec, identicalVec))
  edgeBench.add('chunk_empty', () => chunkTextFallback('', 512))
  edgeBench.add('chunk_long_line_50k', () => chunkTextFallback(longLine, 512))

  await edgeBench.warmup()
  await edgeBench.run()
  console.table(edgeBench.table())

  // ----- Summary -----
  console.log('\n' + '=' .repeat(70))
  console.log('✅ TypeScript benchmarks complete!')
  console.log('\nKey metrics to compare with Rust:')
  console.log('  - cosine_similarity_dim_1024')
  console.log('  - chunk_medium_10kb')
  console.log('  - hybrid_merge_results_1000')
}

runBenchmarks().catch(console.error)

File: services/zero-core/benches/performance.rs
//! Performance benchmarks for zero-core memory operations.
//!
//! Run with: cargo bench -p zero-core
//!
//! SIMD benchmarks require the `simd` feature (enabled by default).

use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion, Throughput};
use zero_core::memory::{
    chunk_markdown, cosine_similarity, dot_product, euclidean_distance, hybrid_merge, normalize,
};

/// Benchmark cosine similarity across different vector dimensions.
fn bench_cosine_similarity(c: &mut Criterion) {
    let mut group = c.benchmark_group("cosine_similarity");

    for size in [128, 512, 1024, 1536, 4096, 16384] {
        group.throughput(Throughput::Elements(size as u64));

        // Generate test vectors
        let a: Vec<f32> = (0..size).map(|i| (i as f32 / size as f32).sin()).collect();
        let b: Vec<f32> = (0..size).map(|i| (i as f32 / size as f32).cos()).collect();

        group.bench_with_input(BenchmarkId::new("dim", size), &size, |bench, _| {
            bench.iter(|| cosine_similarity(&a, &b))
        });
    }
    group.finish();
}

/// Benchmark euclidean distance across different vector dimensions.
fn bench_euclidean_distance(c: &mut Criterion) {
    let mut group = c.benchmark_group("euclidean_distance");

    for size in [128, 512, 1024, 1536, 4096] {
        group.throughput(Throughput::Elements(size as u64));

        let a: Vec<f32> = (0..size).map(|i| (i as f32 / size as f32).sin()).collect();
        let b: Vec<f32> = (0..size).map(|i| (i as f32 / size as f32).cos()).collect();

        group.bench_with_input(BenchmarkId::new("dim", size), &size, |bench, _| {
            bench.iter(|| euclidean_distance(&a, &b))
        });
    }
    group.finish();
}

/// Benchmark dot product across different vector dimensions.
fn bench_dot_product(c: &mut Criterion) {
    let mut group = c.benchmark_group("dot_product");

    for size in [128, 512, 1024, 1536, 4096] {
        group.throughput(Throughput::Elements(size as u64));

        let a: Vec<f32> = (0..size).map(|i| (i as f32 / size as f32).sin()).collect();
        let b: Vec<f32> = (0..size).map(|i| (i as f32 / size as f32).cos()).collect();

        group.bench_with_input(BenchmarkId::new("dim", size), &size, |bench, _| {
            bench.iter(|| dot_product(&a, &b))
        });
    }
    group.finish();
}

/// Benchmark vector normalization across different dimensions.
fn bench_normalize(c: &mut Criterion) {
    let mut group = c.benchmark_group("normalize");

    for size in [128, 512, 1024, 1536, 4096] {
        group.throughput(Throughput::Elements(size as u64));

        let v: Vec<f32> = (0..size).map(|i| (i as f32 / size as f32).sin()).collect();

        group.bench_with_input(BenchmarkId::new("dim", size), &size, |bench, _| {
            bench.iter(|| normalize(&v))
        });
    }
    group.finish()
}

/// Benchmark markdown chunking with varying document sizes.
fn bench_chunk_markdown(c: &mut Criterion) {
    let mut group = c.benchmark_group("chunk_markdown");

    // Small document (~100 chars)
    let small = "# Title\n\nShort content paragraph here.\n\n## Section\n\nMore text.";

    // Medium document (~10KB) - typical code file
    let medium = generate_markdown_document(50);

    // Large document (~100KB) - large documentation file
    let large = generate_markdown_document(500);

    group.bench_function("small_100b", |b| b.iter(|| chunk_markdown(small, 512)));
    group.bench_function("medium_10kb", |b| b.iter(|| chunk_markdown(&medium, 512)));
    group.bench_function("large_100kb", |b| b.iter(|| chunk_markdown(&large, 512)));

    // Also test with different chunk sizes
    group.bench_function("medium_256tok", |b| b.iter(|| chunk_markdown(&medium, 256)));
    group.bench_function("medium_1024tok", |b| b.iter(|| chunk_markdown(&medium, 1024)));

    group.finish();
}

/// Generate a synthetic markdown document with n sections.
fn generate_markdown_document(sections: usize) -> String {
    let mut doc = String::with_capacity(sections * 200);
    doc.push_str("# Document Title\n\n");
    doc.push_str("Introduction paragraph with some content.\n\n");

    for i in 0..sections {
        doc.push_str(&format!("## Section {}\n\n", i + 1));
        doc.push_str("This is a paragraph of text that contains some content. ");
        doc.push_str("It has multiple sentences to simulate real documentation. ");
        doc.push_str(&format!("This is section number {}.\n\n", i + 1));

        // Add code block occasionally
        if i % 5 == 0 {
            doc.push_str("```rust\nfn example() -> u32 {\n    42\n}\n```\n\n");
        }

        // Add list occasionally
        if i % 3 == 0 {
            doc.push_str("- Item one\n- Item two\n- Item three\n\n");
        }
    }

    doc
}

/// Benchmark hybrid merge with varying result set sizes.
fn bench_hybrid_merge(c: &mut Criterion) {
    let mut group = c.benchmark_group("hybrid_merge");

    for size in [50, 100, 500, 1000, 5000, 10000] {
        // Generate vector results (simulating embeddings search)
        let vec_results: Vec<(String, f32)> = (0..size)
            .map(|i| (format!("doc_{i}"), 0.95 - (i as f32 * 0.0001)))
            .collect();

        // Generate keyword results (simulating BM25 search) - typically fewer
        let kw_results: Vec<(String, f32)> = (0..size / 2)
            .map(|i| {
                // Some overlap with vector results, some unique
                let id = if i % 2 == 0 {
                    format!("doc_{i}")
                } else {
                    format!("kw_doc_{i}")
                };
                (id, 15.0 - (i as f32 * 0.01))
            })
            .collect();

        group.throughput(Throughput::Elements(size as u64));

        group.bench_with_input(BenchmarkId::new("results", size), &size, |bench, _| {
            bench.iter(|| hybrid_merge(&vec_results, &kw_results, 0.7, 0.3, 50))
        });
    }

    group.finish();
}

/// Benchmark edge cases and specific scenarios.
fn bench_edge_cases(c: &mut Criterion) {
    let mut group = c.benchmark_group("edge_cases");

    // Empty inputs
    let empty: Vec<f32> = vec![];
    group.bench_function("cosine_empty", |b| b.iter(|| cosine_similarity(&empty, &empty)));

    // Identical vectors (should be ~1.0)
    let identical: Vec<f32> = (0..1024).map(|i| i as f32).collect();
    group.bench_function("cosine_identical_1024", |b| {
        b.iter(|| cosine_similarity(&identical, &identical))
    });

    // Orthogonal vectors
    let ortho_a: Vec<f32> = (0..1024).map(|i| if i < 512 { 1.0 } else { 0.0 }).collect();
    let ortho_b: Vec<f32> = (0..1024).map(|i| if i >= 512 { 1.0 } else { 0.0 }).collect();
    group.bench_function("cosine_orthogonal_1024", |b| {
        b.iter(|| cosine_similarity(&ortho_a, &ortho_b))
    });

    // Empty markdown
    group.bench_function("chunk_empty", |b| b.iter(|| chunk_markdown("", 512)));

    // Single long line (stress test)
    let long_line = "word ".repeat(10000);
    group.bench_function("chunk_long_line_50k", |b| b.iter(|| chunk_markdown(&long_line, 512)));

    // Deep heading nesting
    let deep_headings = (1..50)
        .map(|i| format!("## Section {}\nContent for section {}.\n\n", i, i))
        .collect::<String>();
    group.bench_function("chunk_deep_headings", |b| {
        b.iter(|| chunk_markdown(&deep_headings, 512))
    });

    // Hybrid merge with no overlap
    let vec_only: Vec<(String, f32)> = (0..100).map(|i| (format!("vec_{i}"), 0.9)).collect();
    let kw_only: Vec<(String, f32)> = (0..100).map(|i| (format!("kw_{i}"), 10.0)).collect();
    group.bench_function("hybrid_no_overlap_100", |b| {
        b.iter(|| hybrid_merge(&vec_only, &kw_only, 0.7, 0.3, 50))
    });

    // Hybrid merge with full overlap
    let ids: Vec<String> = (0..100).map(|i| format!("doc_{i}")).collect();
    let vec_full: Vec<(String, f32)> = ids.iter().map(|id| (id.clone(), 0.9)).collect();
    let kw_full: Vec<(String, f32)> = ids.iter().map(|id| (id.clone(), 10.0)).collect();
    group.bench_function("hybrid_full_overlap_100", |b| {
        b.iter(|| hybrid_merge(&vec_full, &kw_full, 0.7, 0.3, 50))
    });

    group.finish();
}

criterion_group!(
    benches,
    bench_cosine_similarity,
    bench_euclidean_distance,
    bench_dot_product,
    bench_normalize,
    bench_chunk_markdown,
    bench_hybrid_merge,
    bench_edge_cases
);
criterion_main!(benches);
